### PR TITLE
Parallelize dmg building

### PR DESCRIPTION
**Pull request by vasi**
Uses parallelism when building a dmg, as [proposed in bugzilla](https://bugzilla.mozilla.org/show_bug.cgi?id=1916320). On my 12-core system, it goes from 30s to 5s.

Tested on Ubuntu 22.04 with:
```
$ curl -Lo orig.dmg 'https://download-installer.cdn.mozilla.net/pub/firefox/releases/130.0/mac/en-CA/Firefox%20130.0.dmg'
$ md5sum orig.dmg
6bcf91f8632dc7751a6736354155750c
$ git checkout mozilla
$ cmake -DCMAKE_BUILD_TYPE=Release
$ make -j12
$ dmg/dmg extract orig.dmg orig.hfs >/dev/null
$ md5sum orig.hfs
e68bcc9b1de3ccf899aa46151df77b52
$ time dmg/dmg build orig.hfs base.dmg __MOZCUSTOM__ > /dev/null
30.25s user 0.22s system 99% cpu 30.481 total
$ git checkout mozilla-parallel
$ make -j12
$ time dmg/dmg build orig.hfs parallel.dmg __MOZCUSTOM__ > /dev/null
47.04s user 1.98s system 867% cpu 5.654 total
$ md5sum base.dmg parallel.dmg
0d2c2af7c3b0dfe2950d5974536bfd33  base.dmg
0d2c2af7c3b0dfe2950d5974536bfd33  parallel.dmg
```

Overall it looks like a big change, but each of the commits is independent, small and relatively self-explanatory (I hope). Some design notes that may be helpful:

* We use N identical threads, which can compress in parallel, but only one thread can be reading or writing at a time. This allows serial tasks such as checksums to remain serialized, as long as they happen at read-time or write-time.
* When we finish compressing a block, we can't write it until any previous blocks are ready, so we keep a linked-list queue of pending blocks.
* I've made a couple of small changes or simplifications as I went:
   * All the `addComment`-related code was completely unused, so I removed it.
   * I've removed the unused CRCZeroes stuff, and replaced CRC32 with a call to zlib. It's faster!
   * I commented out many print statements, `dmg` was getting far too verbose to be useful
   * Instead of referring to `blkx->runs[curRun]` over and over again, I use a local BLKXRun variable 
   * We no longer seek back-and-forth in the input file. If we already read the "next block", we just reuse it as our next "current block"
   * I've disentangled the "should this block be raw?" decision from the writing code. This means one fewer callsite of `out->write()` and `compressedChk()`, which I think is simpler.

cc @bhearsum

### CODE REVIEW

**Comment by bhearsum** [comment]
Thank you for opening this! It may take me a few days to take a deep look at it. If you have a chance in the meantime, could you please try running the tests with cram? Eg:
```
pipx install cram
cram test/*.t
```

(Any of the tests that require `xattr` will only succeed on macOS, so some failures may be expected depending where you're running them.)

**Comment by vasi** [comment]
Hmm, I can't seem to get these tests even passing on the `mozilla` branch. Some of them seem to need some setup which I can guess, but even so:

```
$ uname -a
Darwin Harpy 23.6.0 Darwin Kernel Version 23.6.0: Fri Jul  5 17:54:20 PDT 2024; root:xnu-10063.141.1~2/RELEASE_X86_64 x86_64
$ git checkout mozilla
$ cmake -B build
$ cram test/attribution.t
--- test/attribution.t
+++ test/attribution.t.err
@@ -79,10 +79,10 @@
   @@ -17159,8 +17159,8 @@
    00043060: 0000 0000 0000 0000 0000 0004 064c 0000  .............L..
    00043070: 0000 0000 0000 0000 0000 0000 0000 0000  ................
-   00043080: 0001 0000 0001 7348 3366 6998 3c64 c623  ......sH3fi.<d.#
-  -00043090: 7b32 6745 8b6b 0000 0002 0000 0020 8e77  {2gE.k....... .w
+   00043080: 0001 0000 0001 2a0c b53a d9ac b760 f13a  ......*..:...`.:
+  -00043090: d610 a741 0000 0000 0002 0000 0020 8e77  ...A......... .w
   -000430a0: 24c9 0000 0000 0000 0000 0000 0000 0000  $...............
-  +00043090: 7b32 6745 8b6b 0000 0002 0000 0020 8eb5  {2gE.k....... ..
+  +00043090: d610 a741 0000 0000 0002 0000 0020 8eb5  ...A......... ..
   +000430a0: c555 0000 0000 0000 0000 0000 0000 0000  .U..............
    000430b0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    000430c0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
(more lines of errors)
```

I don't suppose there's some reproducible environment that I could use to run these? Or at least if you could tell me more about your current test env, and I can try to reproduce it.

**Comment by vasi** [comment]
Heh, the particular diff I showed above is due to things like:

```
koly.fUDIFSegmentID.data1 = rand();
```

With the same seed, we should get the same value across runs. But not necessarily across libc versions, or operating systems! If we really want these tests to be reproducible, we'd need to create our own `rand()` implementation. 😢

**Comment by bhearsum** [comment]
Hm...these were definitely passing fully at one point in time. I can reproduce those failures now though :(. I don't have time to work through this at the moment though, so we'll have to live without them. Sorry!

**Comment by bhearsum** [comment]
I tried to use this revision for some of Firefox builds, but unfortunately it failed to compiled :(. https://treeherder.mozilla.org/jobs?repo=try&revision=8de8e171883ddf35d9edf93ca21550b360a6e59a&selectedTaskRun=U7MouFwpQHuAxL5x04iy4g.0

```
[task 2024-09-04T19:02:45.099Z] /builds/worker/fetches/libdmg-hfsplus/dmg/checksum.c:68:20: error: call to undeclared function 'crc32_z'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
[task 2024-09-04T19:02:45.099Z]    68 |   return (*ckSum = crc32_z(*ckSum, buf, len));
[task 2024-09-04T19:02:45.099Z]       |                    ^
[task 2024-09-04T19:02:45.099Z] /builds/worker/fetches/libdmg-hfsplus/dmg/checksum.c:68:20: note: did you mean 'crc32'?
[task 2024-09-04T19:02:45.100Z] /builds/worker/fetches/sysroot/usr/include/zlib.h:1600:23: note: 'crc32' declared here
[task 2024-09-04T19:02:45.100Z]  1600 | ZEXTERN uLong ZEXPORT crc32   OF((uLong crc, const Bytef *buf, uInt len));
[task 2024-09-04T19:02:45.100Z]       |                       ^
[task 2024-09-04T19:02:45.103Z] /builds/worker/fetches/libdmg-hfsplus/dmg/attribution.c:148:122: warning: format specifies type 'unsigned long long' but the argument has type 'int64_t' (aka 'long') [-Wformat]
[task 2024-09-04T19:02:45.103Z]   148 |   printf("data->beforeRaw_UncompressedChkToken CRC32: 0x%x, length: 0x%llx\n", data->beforeRaw_UncompressedChkToken.crc, data->beforeRaw_UncompressedLength);
[task 2024-09-04T19:02:45.103Z]       |                                                                       ~~~~                                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[task 2024-09-04T19:02:45.103Z]       |                                                                       %lx
[task 2024-09-04T19:02:45.103Z] /builds/worker/fetches/libdmg-hfsplus/dmg/attribution.c:149:110: warning: format specifies type 'unsigned long long' but the argument has type 'int64_t' (aka 'long') [-Wformat]
[task 2024-09-04T19:02:45.103Z]   149 |   printf("data->raw_UncompressedChkToken CRC32: 0x%x, length: 0x%llx\n", data->raw_UncompressedChkToken.crc, data->raw_UncompressedLength);
[task 2024-09-04T19:02:45.103Z]       |                                                                 ~~~~                                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
[task 2024-09-04T19:02:45.103Z]       |                                                                 %lx
[task 2024-09-04T19:02:45.103Z] /builds/worker/fetches/libdmg-hfsplus/dmg/attribution.c:150:120: warning: format specifies type 'unsigned long long' but the argument has type 'int64_t' (aka 'long') [-Wformat]
[task 2024-09-04T19:02:45.104Z]   150 |   printf("data->afterRaw_UncompressedChkToken CRC32: 0x%x, length: 0x%llx\n", data->afterRaw_UncompressedChkToken.crc, data->afterRaw_UncompressedLength);
[task 2024-09-04T19:02:45.104Z]       |                                                                      ~~~~                                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[task 2024-09-04T19:02:45.104Z]       |                                                                      %lx
[task 2024-09-04T19:02:45.104Z] /builds/worker/fetches/libdmg-hfsplus/dmg/attribution.c:157:118: warning: format specifies type 'unsigned long long' but the argument has type 'int64_t' (aka 'long') [-Wformat]
[task 2024-09-04T19:02:45.104Z]   157 |   printf("data->beforeRaw_CompressedChkToken CRC32: 0x%x, length: 0x%llx\n", data->beforeRaw_CompressedChkToken.crc, data->beforeRaw_CompressedLength);
[task 2024-09-04T19:02:45.104Z]       |                                                                     ~~~~                                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[task 2024-09-04T19:02:45.104Z]       |                                                                     %lx
[task 2024-09-04T19:02:45.104Z] /builds/worker/fetches/libdmg-hfsplus/dmg/attribution.c:158:106: warning: format specifies type 'unsigned long long' but the argument has type 'int64_t' (aka 'long') [-Wformat]
[task 2024-09-04T19:02:45.104Z]   158 |   printf("data->raw_CompressedChkToken CRC32: 0x%x, length: 0x%llx\n", data->raw_CompressedChkToken.crc, data->raw_CompressedLength);
[task 2024-09-04T19:02:45.104Z]       |                                                               ~~~~                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~
[task 2024-09-04T19:02:45.104Z]       |                                                               %lx
[task 2024-09-04T19:02:45.104Z] /builds/worker/fetches/libdmg-hfsplus/dmg/attribution.c:159:116: warning: format specifies type 'unsigned long long' but the argument has type 'int64_t' (aka 'long') [-Wformat]
[task 2024-09-04T19:02:45.104Z]   159 |   printf("data->afterRaw_CompressedChkToken CRC32: 0x%x, length: 0x%llx\n", data->afterRaw_CompressedChkToken.crc, data->afterRaw_CompressedLength);
[task 2024-09-04T19:02:45.104Z]       |                                                                    ~~~~                                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[task 2024-09-04T19:02:45.104Z]       |                                                                    %lx
[task 2024-09-04T19:02:45.105Z] /builds/worker/fetches/libdmg-hfsplus/dmg/attribution.c:183:87: warning: format specifies type 'unsigned long long' but the argument has type 'uint64_t' (aka 'unsigned long') [-Wformat]
[task 2024-09-04T19:02:45.105Z]   183 |   printf("sentinelAfterMainBlkx: %d, 0x%llx, 0x%llx\n", attributionResource->version, attributionResource->rawPos, attributionResource->rawLength);
[task 2024-09-04T19:02:45.105Z]       |                                        ~~~~                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~
[task 2024-09-04T19:02:45.105Z]       |                                        %lx
[task 2024-09-04T19:02:45.105Z] /builds/worker/fetches/libdmg-hfsplus/dmg/attribution.c:183:116: warning: format specifies type 'unsigned long long' but the argument has type 'uint64_t' (aka 'unsigned long') [-Wformat]
[task 2024-09-04T19:02:45.105Z]   183 |   printf("sentinelAfterMainBlkx: %d, 0x%llx, 0x%llx\n", attributionResource->version, attributionResource->rawPos, attributionResource->rawLength);
[task 2024-09-04T19:02:45.105Z]       |                                                ~~~~                                                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[task 2024-09-04T19:02:45.105Z]       |                                                %lx
[task 2024-09-04T19:02:45.106Z] /builds/worker/fetches/libdmg-hfsplus/dmg/attribution.c:248:10: warning: format specifies type 'unsigned long long' but the argument has type 'uint64_t' (aka 'unsigned long') [-Wformat]
[task 2024-09-04T19:02:45.106Z]   247 |   printf("attribution: at 0x%llx, 0x%llx bytes\n",
[task 2024-09-04T19:02:45.106Z]       |                             ~~~~
[task 2024-09-04T19:02:45.106Z]       |                             %lx
[task 2024-09-04T19:02:45.106Z]   248 |          attributionResource->rawPos,
[task 2024-09-04T19:02:45.106Z]       |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~
[task 2024-09-04T19:02:45.106Z] /builds/worker/fetches/libdmg-hfsplus/dmg/attribution.c:249:10: warning: format specifies type 'unsigned long long' but the argument has type 'uint64_t' (aka 'unsigned long') [-Wformat]
[task 2024-09-04T19:02:45.106Z]   247 |   printf("attribution: at 0x%llx, 0x%llx bytes\n",
[task 2024-09-04T19:02:45.106Z]       |                                     ~~~~
[task 2024-09-04T19:02:45.106Z]       |                                     %lx
[task 2024-09-04T19:02:45.106Z]   248 |          attributionResource->rawPos,
[task 2024-09-04T19:02:45.106Z]   249 |          attributionResource->rawLength);
[task 2024-09-04T19:02:45.106Z]       |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[task 2024-09-04T19:02:45.107Z] /builds/worker/fetches/libdmg-hfsplus/dmg/attribution.c:274:36: warning: format specifies type 'unsigned long long' but the argument has type 'int64_t' (aka 'long') [-Wformat]
[task 2024-09-04T19:02:45.107Z]   274 |   printf("anchorOffset: 0x%llx\n", anchorOffset);
[task 2024-09-04T19:02:45.107Z]       |                           ~~~~     ^~~~~~~~~~~~
[task 2024-09-04T19:02:45.107Z]       |                           %lx
[task 2024-09-04T19:02:45.109Z] 1 error generated.
[task 2024-09-04T19:02:45.111Z] make[2]: *** [dmg/CMakeFiles/dmg.dir/build.make:104: dmg/CMakeFiles/dmg.dir/checksum.c.o] Error 1
[task 2024-09-04T19:02:45.111Z] make[2]: *** Waiting for unfinished jobs....
[task 2024-09-04T19:02:45.114Z] 11 warnings generated.
[task 2024-09-04T19:02:45.116Z] /builds/worker/fetches/libdmg-hfsplus/dmg/resources.c:398:13: warning: assigning to 'char *' from 'unsigned char *' converts between pointers to integer types where one is of the unique plain 'char' type and the other is not [-Wpointer-sign]
[task 2024-09-04T19:02:45.116Z]   398 |         data->name = attributionFromName;
[task 2024-09-04T19:02:45.116Z]       |                    ^ ~~~~~~~~~~~~~~~~~~~
[task 2024-09-04T19:02:45.116Z] /builds/worker/fetches/libdmg-hfsplus/dmg/resources.c:399:26: warning: passing 'char *' to parameter of type 'unsigned char *' converts between pointers to integer types where one is of the unique plain 'char' type and the other is not [-Wpointer-sign]
[task 2024-09-04T19:02:45.116Z]   399 |         flipAttributionResource(data->name, 0);
[task 2024-09-04T19:02:45.116Z]       |                                 ^~~~~~~~~~
[task 2024-09-04T19:02:45.116Z] /builds/worker/fetches/libdmg-hfsplus/dmg/resources.c:125:52: note: passing argument to parameter 'data' here
[task 2024-09-04T19:02:45.116Z]   125 | static void flipAttributionResource(unsigned char* data, char out) {
[task 2024-09-04T19:02:45.116Z]       |                                                    ^
[task 2024-09-04T19:02:45.122Z] /builds/worker/fetches/libdmg-hfsplus/dmg/resources.c:917:19: warning: assigning to 'char *' from 'unsigned char *' converts between pointers to integer types where one is of the unique plain 'char' type and the other is not [-Wpointer-sign]
[task 2024-09-04T19:02:45.122Z]   917 |     curData->name = (unsigned char*) malloc(nameLength);
[task 2024-09-04T19:02:45.122Z]       |                   ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[task 2024-09-04T19:02:45.134Z] 3 warnings generated.
[task 2024-09-04T19:02:45.138Z] [ 84%] Linking C executable hfsplus
[task 2024-09-04T19:02:45.139Z] cd /builds/worker/fetches/libdmg-hfsplus/hfs && /usr/bin/cmake -E cmake_link_script CMakeFiles/hfsplus.dir/link.txt --verbose=1
[task 2024-09-04T19:02:45.140Z] make[2]: Leaving directory '/builds/worker/fetches/libdmg-hfsplus'
[task 2024-09-04T19:02:45.140Z] make[1]: *** [CMakeFiles/Makefile2:183: dmg/CMakeFiles/dmg.dir/all] Error 2
[task 2024-09-04T19:02:45.140Z] make[1]: *** Waiting for unfinished jobs....
```

(The build script being used is https://searchfox.org/mozilla-central/source/taskcluster/scripts/misc/build-dmg-hfsplus.sh, and it's being compiled in [this docker image](https://firefox-ci-tc.services.mozilla.com/api/queue/v1/task/ZIdnRuOkTICzRJugz9BOzg/runs/0/artifacts/public%2Fimage.tar.zst))

**Comment by vasi** [comment]
~~Thanks! Do you know where I can get the toolchain/sysroot artifacts from?~~
Never mind, there are URLs in the [logs](https://treeherder.mozilla.org/logviewer?job_id=472938196&repo=try&lineNumber=352).

**Comment by vasi** [comment]
Ok, pushed fixes for the compile errors.

**Comment by vasi** [comment]
Opened #10 to look into test reproducibility.

**Comment by bhearsum** [comment]
> ~Thanks! Do you know where I can get the toolchain/sysroot artifacts from?~ Never mind, there are URLs in the [logs](https://treeherder.mozilla.org/logviewer?job_id=472938196&repo=try&lineNumber=352).

You can probably use your own tools there; I think the main thing is just compiling with the same flags. But if you've already downloaded the ones we're using, that's fine too :).

I've thrown the latest version of this patch at our CI as well: https://treeherder.mozilla.org/jobs?repo=try&revision=e8c51d3f14101cf01db435c643225863c96fdf82. We'll see if it compiles, and hopefully get an attributed DMG out of it as an additional sanity check.

**Comment by bhearsum** [comment]
> > ~Thanks! Do you know where I can get the toolchain/sysroot artifacts from?~ Never mind, there are URLs in the [logs](https://treeherder.mozilla.org/logviewer?job_id=472938196&repo=try&lineNumber=352).
> 
> You can probably use your own tools there; I think the main thing is just compiling with the same flags. But if you've already downloaded the ones we're using, that's fine too :).
> 
> I've thrown the latest version of this patch at our CI as well: https://treeherder.mozilla.org/jobs?repo=try&revision=e8c51d3f14101cf01db435c643225863c96fdf82. We'll see if it compiles, and hopefully get an attributed DMG out of it as an additional sanity check.

The push above ended up suffering from some sort of freak issue causing a startup crash (even the non-attributed DMG, which rules out this change being a factor).

I repushed in https://treeherder.mozilla.org/jobs?repo=try&revision=32d33f4f4823b6a703573d920680483390167496, and [that attributed dmg](https://firefox-ci-tc.services.mozilla.com/api/queue/v1/task/cC1PSJeFR2aC0wBeylPhSA/runs/0/artifacts/public/build/target.dmg) seems to be working fine.

Comparing the DMG build times shows a positive result too. The two pushes above took 22 and 21 seconds respectively to build their DMGs. A [comparable job on mozilla-central](https://treeherder.mozilla.org/jobs?repo=mozilla-central&searchStr=repackage-macosx64-shippable%2Fopt&selectedTaskRun=CW3h99p5Sbm9UsqA1rA3Lg.0) took 1m11s - quite the speedup! (Note that these jobs produce other package types as well, so you have to dig into the logs to see just the dmg times.)

**Comment by vasi** [comment]
Hey that's great to see! Hopefully that means less time for Mozilla devs to spend waiting on CI :)

**Comment by vasi** [comment]
Rebased on #13 , and tests pass: https://github.com/vasi/libdmg-hfsplus/actions/runs/10907052141/job/30269776105

**Comment by bhearsum** [comment]
Between the tests here and in Firefox's CI I'm fairly confident in this. I'll need to find some time to have a deeper read of this patch before I approve and merge though - hopefully this will be in the next week or two.

**Comment by vasi** [comment]
That's reasonable! Note that it's almost certainly easier to review commit-by-commit than all at once, I tried to make them all look fairly small and obvious. 

Step-by-step, the process looks like:

1. Turn local variables into a `threadData` struct, so we can have multiple functions instead of one giant one.
2. Remove cruft to simplify code: unnecessary seeks, `addComment` stuff, logging.
3. Extract the concept of a "block of data", so we can think about passing blocks between functions.
4. Separate input/compression/output into functions. Each function should have ~exclusive access to certain fields of our struct, to prevent races.
5. Start running our functions in multiple threads, and boom we're parallel.

Let me know if you have any questions, I'm happy to explain anything that's unclear.

**Comment by vasi** [comment]
@bhearsum, have you had time to take a look at this?

**Comment by bhearsum** [comment]
> @bhearsum, have you had time to take a look at this?

Not in the depth needed :(. I haven't forgotten about it though.

**Review by bhearsum — APPROVED** [review]
I finally had a chance to dive into this - apologies for the delay!

This looks great to me, and it appears to avoid compressing blocks that are going to be kept raw anyways ([which I noticed currently happens](https://github.com/mozilla/libdmg-hfsplus/blob/07b845168e88716fd0248c10126b2971e11bea80/dmg/io.c#L251-L257)).

The only thing slightly concerning is the removal of `addComment` support. I do see that we never seem to actually use that, so it ought to be fine. However, we do use this for [ios builds](https://searchfox.org/mozilla-central/rev/0d0eef18ea91e38ddfb82a1ac5d6560a146da5c0/taskcluster/kinds/build/ios.yml#32), so I'm doing [some additional testing](https://treeherder.mozilla.org/jobs?repo=try&revision=1ac6dc555c208781c1319ab99c0547df475a1e41) with that on the off chance I'm misreading something - but I don't anticipate that catching anything.

Once that test is done I'll get this merged. I intend to pick up the new revision near the beginning of a Firefox Nightly cycle - most likely the last week of November.

Thank you so much again for this contribution and your patience - this will have an impact on build and repackage jobs, which we run thousands of daily!

**Comment by vasi** [comment]
Thank you so much for the review! I know it's a lot of work to look over such a large PR, especially one you didn't expect.

**Comment by vasi** [comment]
Just an FYI, there's a couple of other changes I'm considering making:
1. Allowing the user to specify a compression algorithm and/or compression level. This would allow choosing lzfse (or gzip) for higher speed, which could be useful for dev builds. Or, LZMA2 for even higher compression ratios.
2. Adding support for streaming input, which I'd use for converting compressed-raw-images to compressed-dmg.

Change 1 is plausibly useful to Mozilla, but I'm not sure. Change 2 probably isn't, but might be useful to other users of libdmg. Do you have any preferences on how I do this? I could either submit them as PRs, or maintain a long-term fork, whatever works best for y'all.

**Comment by bhearsum** [comment]
PRs would probably be fine; I agree that the second one is unlikely to be desirable for us.

**Pull request merged**
